### ISSUE TITLE
About the lacking of query sample in data_loader and the lacking of target in calculating loss

### ISSUE THREAD

**Issue by quanghuy0497** [wyharveychen/CloserLookFewShot#74]
Hi Harvey, 

I am currently utilizing your code for my project and I find these two issues, can you help me answer them?
1. **The difference in query size between 1-shot and 5-shot**:
    + Normally, the training batch in train_loader, val_loader has the form **[n_way, n_support + n_query, dim, h, w]**. With the 5-way setting, query size = 16 per class, training on the Omniglot dataset, the data batch should have the shape `[5, n_support + 16, 3, 28, 28]`
    + In the case that I use the ***1-shot setting***, the training/testing batch has the shape `[5, 17, 3, 28, 28]` and it's completely normal. 
    + However, when I use the ***5-shot setting***, the training/testing batch has the shape `[5, 20, 3, 28, 28]` while the second axis size should be 21 = 5 + 16. 
    + When I trace back to the data loader generator in **datamgr.py** `class SetDataManager`, the `self.batch_size` variable is still equal to 21 in the line `dataset = SetDataset( data_file , self.batch_size, transform )`. However, the generated data_loader (the last line of `datamgr.py` has a batch size of 20.
   + Back to **meta_template.py** ` def train_loop`, the query size is 15 by the code `self.n_query = x.size(1) - self.n_support` and thus there are only 15 query samples per class.
    + So, does losing 1 query sample/class affects the training/testing performance? Is it a fair comparison between 1-shot and 5-shot without 2 difference batch_size?

2. **The lacking of target value for calculating loss**:
    + In the **meta_template.py**, I find that the `def train_loop` only uses x for training (`for i, (x,_ ) in enumerate(train_loader)`) and calculating loss only with x `loss = self.set_forward_loss(x)`. Thus, there is no ground-truth value (let's call it `y_query_target`) to compare with the predicted x inside the loss function.
    + Inside the particular methods (relationnet, protonet,...) I find that the y_query (let's call it `y_query_generated`) is generated by the command `y_query = Variable(torch.from_numpy(np.repeat(range(self.n_way), self.n_query) )).cuda()` inside the function **def set_forward_loss(self, x)**. 
    + Therefore, I wonder how the loss is calculated without the ground-truth value y_query. Does the `y_query_genrated` equal to the `y_query_target`?

I would really appreciate it if anyone can help me with these two issues above? Thank you for your time.

Yours faithfully,
